analysis/tests/src/expected/Completion.res.txt

```diff
@@ -1381,3 +1383,51 @@ posCursor:[355:23] posNoWhite:[355:22] Found expr:[0:-1->355:23]
 posCursor:[355:23] posNoWhite:[355:22] Found expr:[355:12->355:23]
 []
 
+Complete tests/src/Completion.res 362:8
+posCursor:[362:8] posNoWhite:[362:7] Found expr:[360:8->365:3]
+posCursor:[362:8] posNoWhite:[362:7] Found expr:[361:2->365:3]
+posCursor:[362:8] posNoWhite:[362:7] Found pattern:[362:7->364:5]
+posCursor:[362:8] posNoWhite:[362:7] Found pattern:[362:7->362:8]
+XXX Ppat_construct T:[362:7->362:8]
+Completable: Cpath Value[T]
+[{
+    "label": "That",
+    "kind": 4,
+    "tags": [],
+    "detail": "That\n\ntype v =  This | That",
+    "documentation": null
+  }, {
+    "label": "This",
+    "kind": 4,
+    "tags": [],
+    "detail": "This\n\ntype v =  This | That",
+    "documentation": null
+  }, {
+    "label": "TableclothMap",
```

Comment:
This is a module starting with `T` which happens to be in scope.
This is accurate, as the user might be in the process of writing `ThatSpecificModule.SomeConstructor`.

Reply:
I disagree with this, except in the rare case where the type of the switch variable is not known (because it might be used to disambiguate, as you suggested). That should only happen once, or in many cases never.

My understanding is that once the type is known the compiler no longer needs disambiguation to find constructors. At that point seeing all scope modules in the autocomplete list is just noise.

https://rescript-lang.org/try?code=LYewJgrgNgpgBAeQC4AsYCc4F44G8BQccSAngA7wBu2iAdvAD5wAqA7iHE8yujDIXFhI4AMxqUAhlAjwsAPgFEAzqwCWSAMYo4k6fAJEiAeiNxaHMKqUTgAI1UBzCBKSqQtRZzqy5cAIyeXOzYvgBMgSw8fCFwAMyeAL74SfhCouJSMiECKupaOpn6AiZwSqq0DrBwltZ2js6u7gJMyGjoAHQI9DEBREEc8nDhfZG8PnECSUA

I'm about to start looking at how to remove modules from the autocomplete list in this scenario, but now I realise it was intentional... can we discuss whether removing them is a good idea?

Reply:
I think you're interested in type contexts, here: https://github.com/rescript-lang/rescript-vscode/issues/494
Not this PR.

Reply:
I am, but then I found this PR and it looked like completion in switch statements already included all the valid constructors.

Reply:
This is an ongoing process. Not everything is at the level it could be.
In the process, some techniques have been used to aid autocompletion where the alternative would be to get nothing.